modules/mcc/src/color.cpp

```diff
@@ -0,0 +1,398 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Longbu Wang <wanglongbu@huawei.com.com>
+//         Jinheng Zhang <zhangjinheng1@huawei.com>
+//         Chenqi Shan <shanchenqi@huawei.com>
+
+#include "color.hpp"
+
+namespace cv {
+namespace ccm {
+Color::Color()
+    : colors(Mat())
+    , cs(*new ColorSpace())
+{}
+Color::Color(Mat colors_, enum COLOR_SPACE cs_)
+    : colors(colors_)
+    , cs(*GetCS::get_cs(cs_))
+{}
+
+Color::Color(Mat colors_, const ColorSpace& cs_, Mat colored_)
+    : colors(colors_)
+    , cs(cs_)
+    , colored(colored_)
+{
+    grays = ~colored;
+}
+Color::Color(Mat colors_, enum COLOR_SPACE cs_, Mat colored_)
+    : colors(colors_)
+    , cs(*GetCS::get_cs(cs_))
+    , colored(colored_)
+{
+    grays = ~colored;
+}
+
+Color::Color(Mat colors_, const ColorSpace& cs_)
+    : colors(colors_)
+    , cs(cs_)
+{}
+
+Color Color::to(const ColorSpace& other, CAM method, bool save)
+{
+    if (history.count(other) == 1)
+    {
+        return *history[other];
+    }
+    if (cs.relate(other))
+    {
+        return Color(cs.relation(other).run(colors), other);
+    }
+    Operations ops;
+    ops.add(cs.to).add(XYZ(cs.io).cam(other.io, method)).add(other.from);
+    std::shared_ptr<Color> color(new Color(ops.run(colors), other));
+    if (save)
+    {
+        history[other] = color;
+    }
+    return *color;
+}
+Color Color::to(COLOR_SPACE other, CAM method, bool save)
+{
+    return to(*GetCS::get_cs(other), method, save);
+}
+
+Mat Color::channel(Mat m, int i)
+{
+    Mat dchannels[3];
+    split(m, dchannels);
+    return dchannels[i];
+}
+
+Mat Color::toGray(IO io, CAM method, bool save)
+{
+    XYZ xyz = *XYZ::get(io);
+    return channel(this->to(xyz, method, save).colors, 1);
+}
+
+Mat Color::toLuminant(IO io, CAM method, bool save)
+{
+    Lab lab = *Lab::get(io);
+    return channel(this->to(lab, method, save).colors, 0);
+}
+
+Mat Color::diff(Color& other, DISTANCE_TYPE method)
+{
+    return diff(other, cs.io, method);
+}
+
+Mat Color::diff(Color& other, IO io, DISTANCE_TYPE method)
+{
+    Lab lab = *Lab::get(io);
+    switch (method)
+    {
+    case cv::ccm::CIE76:
+    case cv::ccm::CIE94_GRAPHIC_ARTS:
+    case cv::ccm::CIE94_TEXTILES:
+    case cv::ccm::CIE2000:
+    case cv::ccm::CMC_1TO1:
+    case cv::ccm::CMC_2TO1:
+        return distance(to(lab).colors, other.to(lab).colors, method);
+    case cv::ccm::RGB:
+        return distance(to(*cs.nl).colors, other.to(*cs.nl).colors, method);
+    case cv::ccm::RGBL:
+        return distance(to(*cs.l).colors, other.to(*cs.l).colors, method);
+    default:
+        throw std::invalid_argument { "Wrong method!" };
```

Comment:
Please replace all "throw" statements to `CV_Error`:
- `CV_Error(Error::StsBadArg, "message")`
- `CV_Error(Error::StsNotImplemented, "")`

Reply:
OK, thanks. I fixed this issue.